src/vm/proftoeeinterfaceimpl.cpp

```diff
@@ -6372,6 +6379,320 @@ HRESULT ProfToEEInterfaceImpl::GetFunctionInfo2(FunctionID funcId,
 }
 
 /*
+* FunctionHasNoMetadata
+*
+* This function takes a functionId that maybe of a metadata-less method like an IL Stub
+* or LCG method and returns true in the pHasNoMetadata if it is indeed a metadata-less
+* method.
+*
+* Parameters:
+*   functionId - The function that is being requested.
+*   pHasNoMetadata - An optional parameter for returning if the function has metadata or not.
+*
+* Returns:
+*   S_OK if successful.
+*/
+HRESULT ProfToEEInterfaceImpl::FunctionHasNoMetadata(FunctionID functionId, BOOL *pHasNoMetadata)
+{
+    CONTRACTL
+    {
+    NOTHROW;
+    GC_NOTRIGGER;
+    MODE_ANY;
+    EE_THREAD_NOT_REQUIRED;
+
+    // Generics::GetExactInstantiationsOfMethodAndItsClassFromCallInformation eventually
+    // reads metadata which causes us to take a reader lock.  However, see
+    // code:#DisableLockOnAsyncCalls
+    DISABLED(CAN_TAKE_LOCK);
+
+    // Asynchronous functions can be called at arbitrary times when runtime 
+    // is holding locks that cannot be reentered without causing deadlock.
+    // This contract detects any attempts to reenter locks held at the time 
+    // this function was called.  
+    CANNOT_RETAKE_LOCK;
+
+    SO_NOT_MAINLINE;
+
+    PRECONDITION(CheckPointer(pHasNoMetadata, NULL_OK));
+    }
+    CONTRACTL_END;
+
+    // See code:#DisableLockOnAsyncCalls
+    PERMANENT_CONTRACT_VIOLATION(TakesLockViolation, ReasonProfilerAsyncCannotRetakeLock);
+
+    PROFILER_TO_CLR_ENTRYPOINT_ASYNC_EX(kP2EEAllowableAfterAttach,
+        (LF_CORPROF,
+            LL_INFO1000,
+            "**PROF: FunctionHasNoMetadata 0x%p.\n",
+            functionId));
+
+    //
+    // Verify parameters.
+    //
+
+    if (functionId == NULL)
+    {
+        return E_INVALIDARG;
+    }
+
+    MethodDesc *pMethDesc = FunctionIdToMethodDesc(functionId);
+
+    if (pMethDesc == NULL)
+    {
+        return E_INVALIDARG;
+    }
+
+    // it's not safe to examine a methoddesc that has not been restored so do not do so
+    if (!pMethDesc->IsRestored())
+        return CORPROF_E_DATAINCOMPLETE;
+
+    //
+    // Fill in the pHasNoMetadata, if desired.
+    //
+    if (pHasNoMetadata != NULL)
+    {
+        *pHasNoMetadata = pMethDesc->IsNoMetadata();
+    }
+
+    return S_OK;
+}
+
+/*
+* GetNoMetadataFunctionFromIP
+*
+* This function takes an IP and determines if it is a managed function returning its
+* FunctionID. This method is different from GetFunctionFromIP in that will return
+* FunctionIDs even if they have no associated metadata.
+*
+* Parameters:
+*   ip - The instruction pointer.
+*   pFunctionId - An optional parameter for returning the FunctionID.
+*
+* Returns:
+*   S_OK if successful.
+*/
+HRESULT ProfToEEInterfaceImpl::GetNoMetadataFunctionFromIP(LPCBYTE ip, FunctionID * pFunctionId)
```

Comment:
If we are going to add another iteration of this function, I think we should follow the existing numbering precedent. There is already ICorProfilerInfo4::GetFunctionFromIP2, so this would be GetFunctionFromIP3 and have a super-set of that functionality.